src/pull_request.py

```diff
@@ -0,0 +1,237 @@
+ACTIONS_PULL_REQUEST_BODY = (
+    "This pull request was autogenerated by upload-charm-docs to migrate "
+    "existing documentation from server to the git repository."
+)
+PR_LINK_NO_CHANGE = "<not created due to no changes in repository>"
+BRANCH_PREFIX = "upload-charm-docs"
+DEFAULT_BRANCH_NAME = f"{BRANCH_PREFIX}/migrate"
+
+
+class RepositoryClient:
+    """Wrapper for git/git-server related functionalities."""
+
+    def __init__(self, repository: Repo, github_repository: Repository) -> None:
+        """Construct.
+
+        Args:
+            repository: Client for interacting with local git repository.
+            github_repository: Client for interacting with remote github repository.
+        """
+        self._git_repo = repository
+        self._github_repo = github_repository
+        self._configure_git_user()
+
+    def _configure_git_user(self) -> None:
+        """Configure action git profile defaults.
+
+        Configured profile appears as the git committer.
+        """
+        config_writer = self._git_repo.config_writer()
+        config_writer.set_value("user", "name", ACTIONS_USER_NAME)
+        config_writer.set_value("user", "email", ACTIONS_USER_EMAIL)
+        # there is no context manager, config writer must be manually released.
+        config_writer.release()
+
+    def check_branch_exists(self, branch_name: str) -> bool:
+        """Check if branch exists on remote.
+
+        Args:
+            branch_name: Branch name to check on remote.
+
+        Raises:
+            RepositoryClientError: if unexpected error occurred during git operation.
+
+        Returns:
+            True if branch already exists, False otherwise.
+        """
+        try:
+            self._git_repo.git.fetch("origin", branch_name)
+            return True
+        except GitCommandError as exc:
+            if "couldn't find remote ref" in exc.stderr:
+                return False
+            raise RepositoryClientError(
+                f"Unexpected error checking existing branch. {exc=!r}"
+            ) from exc
+
+    def create_branch(self, branch_name: str, commit_msg: str) -> None:
+        """Create new branch with existing changes.
+
+        Args:
+            branch_name: New branch name.
+            commit_msg: Commit message for current changes.
+
+        Raises:
+            RepositoryClientError: if unexpected error occurred during git operation.
+        """
+        logging.info("create new branch %s", branch_name)
+        try:
+            self._git_repo.git.checkout("-b", branch_name)
+            self._git_repo.git.add(".")
+            self._git_repo.git.commit("-m", f"'{commit_msg}'")
+            self._git_repo.git.push("-u", "origin", branch_name)
+        except GitCommandError as exc:
+            raise RepositoryClientError(f"Unexpected error creating new branch. {exc=!r}") from exc
+
+    def create_github_pull_request(self, branch_name: str, base: str) -> str:
```

Comment:
Should `github` be removed from this since the class is intended to abstract that detail away?